src/charm.py

```diff
@@ -417,10 +419,14 @@ def _on_config_changed(self, event: ConfigChangedEvent) -> None:
         self.render_pgb_config()
         try:
             if self.check_pgb_running():
-                self.reload_pgbouncer()
+                self.reload_pgbouncer(restart=port_changed)
         except ConnectionError:
             event.defer()
 
+        if self.unit.is_leader() and port_changed:
+            # Only update the config once the services have been restarted
+            self.peers.app_databag["current_port"] = str(self.config["listen_port"])
+
```

Comment:
Waiting to the end to set `current_port` to make sure that restarts will happen if the event got deferred.